plugins/modules/cargo.py

```diff
@@ -139,15 +169,29 @@ def get_installed(self):
         cmd = ["install", "--list"]
         data, dummy = self._exec(cmd, True, False, False)
 
-        package_regex = re.compile(r"^([\w\-]+) v(.+):$")
+        package_regex = re.compile(r"^([\w\-]+) v(.+?)(?: \(registry `(.+)`\))?:$")
+
         installed = {}
         for line in data.splitlines():
-            package_info = package_regex.match(line)
-            if package_info:
-                installed[package_info.group(1)] = package_info.group(2)
+            package_match = package_regex.match(line)
+            if package_match:
+                package_name = package_match.group(1)
+                package_info = PackageInfo(
+                    version=package_match.group(2),
+                    registry_url=package_match.group(3),
+                )
+                installed[package_name] = package_info
 
         return installed
 
+    def get_registry_url_from_name(self, name):
+        cargo_config = toml.load_toml(self.config_path)
```

Comment:
I don't see any code that handles the exception thrown in case no TOML library is available.

I guess it's best to define a specific exception type for this in toml.py, and use that here.

(Also please use `from ansible.module_utils.basic import missing_required_lib` and `module.fail_json(msg=missing_required_lib('tomli or toml'))` to fail in that case.)